src/util/tpool/fd_tpool.h

```diff
@@ -0,0 +1,932 @@
+   performance contexts, does no input argument checking.  Specifically,
+   assumes tpool is valid, worker_idx in (0,worker_cnt) (yes, open on
+   both ends), caller is not tpool thread worker_idx, task is valid.
+   worker_idx 0 is special and is considered to always be in the EXEC
+   state so we cannot call exec on it. */
+
+static inline void
+fd_tpool_exec( fd_tpool_t *    tpool,       ulong worker_idx,
+               fd_tpool_task_t task,
+               void *          task_tpool,
+               ulong           task_t0,     ulong task_t1,
+               void *          task_args,
+               void *          task_reduce, ulong task_stride,
+               ulong           task_l0,     ulong task_l1,
+               ulong           task_m0,     ulong task_m1,
+               ulong           task_n0,     ulong task_n1 ) {
+  fd_tpool_private_worker_t * worker = fd_tpool_private_worker( tpool )[ worker_idx ];
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( worker->task        ) = task;       
+  FD_VOLATILE( worker->task_tpool  ) = task_tpool;
+  FD_VOLATILE( worker->task_t0     ) = task_t0;     FD_VOLATILE( worker->task_t1     ) = task_t1;
+  FD_VOLATILE( worker->task_args   ) = task_args;
+  FD_VOLATILE( worker->task_reduce ) = task_reduce; FD_VOLATILE( worker->task_stride ) = task_stride;
+  FD_VOLATILE( worker->task_l0     ) = task_l0;     FD_VOLATILE( worker->task_l1     ) = task_l1;
+  FD_VOLATILE( worker->task_m0     ) = task_m0;     FD_VOLATILE( worker->task_m1     ) = task_m1;
+  FD_VOLATILE( worker->task_n0     ) = task_n0;     FD_VOLATILE( worker->task_n1     ) = task_n1;
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( worker->state ) = FD_TPOOL_WORKER_STATE_EXEC;
+  FD_COMPILER_MFENCE();
+}
+
+/* fd_tpool_waits waits for the tpool thread worker_idx to leave the
+   EXEC state.  As this is used in high performance contexts, does no
+   input argument checking.  Specifically, assumes tpool is valid,
+   worker_idx in (0,worker_cnt) (yes, open on both ends) and caller is
+   not tpool thread worker_idx.  worker_idx 0 is considered to always be
+   in an exec state so we cannot call wait on it. */
+
+static inline void
+fd_tpool_wait( fd_tpool_t const * tpool,
+               ulong              worker_idx ) {
+  int volatile * vstate = (int volatile *)&(fd_tpool_private_worker( tpool )[ worker_idx ]->state);
+  int            state;
+  for(;;) {
+    state = *vstate;
+    if( FD_LIKELY( state!=FD_TPOOL_WORKER_STATE_EXEC ) ) break;
+    FD_SPIN_PAUSE();
+  }
+}
+
+/* Assuming the tasks can be executed safely in any order and/or
+   concurrently, fd_tpool_exec_all_rrobin, fd_tpool_exec_all_block and
+   fd_tpool_exec_all_taskq are functionally equivalent to:
+
+     for( ulong l=l0; l<l1; l++ )
+       task( task_tpool,t0,t1, task_args,task_reduce,task_stride, task_l0,task_l1, l,l+1, t,t+1 );
+
+   where t above is in [t0,t1) and gives the tpool worker thread idx
+   assigned the particular task (and thread t0 is the thread that did
```

Comment:
It took me several times of reading this to understand what it was saying, but now that I get it, I'm not sure how to clarify it. What do you think of this? 
```suggestion
   fd_tpool_exec_all_taskq are functionally equivalent to:

     for( ulong m=m0; m<m1; m++ )
       task( task_tpool,t0,t1, task_args,task_reduce,task_stride, task_l0,task_l1, m,m+1, t,t+1 );

   where t indicates to which tpool worker thread idx the particular 
   task was assigned (and thus t in [t0, t1), where thread t0 is the thread that did
```
I couldn't get out of the 3-d iteration space partitioning (biased from the matrix multiplication example I guess), but my current understanding is these do a 1-d partitioning (partitioning by the second dimension). They pass the same first dimension as a convenience, and they overload the third dimension to tell you which thread, which is also a nice convenience.

Reply:
Done